parser.rst

```diff
@@ -0,0 +1,884 @@
+.. _parser:
+
+How to use CPython's Parser
+===========================
+
+:Author: Pablo Galindo Salgado
+
+.. highlight:: none
+
+Abstract
+--------
+
+The Parser in CPython is currently a `PEG (Parser Expression Grammar)
+<https://en.wikipedia.org/wiki/Parsing_expression_grammar>`_ parser.  The first
+version of the parser used to be an `LL(1)
+<https://en.wikipedia.org/wiki/LL_parser>`_ based parser that was one of the
+oldest parts of CPython implemented before it was replaced by :pep:`617`. In
+particular, both the current parser and the old LL(1) parser are the output of a
+`parser generator <https://en.wikipedia.org/wiki/Compiler-compiler>`_. This
+means that the way the parser is written is by feeding a description of the
+Grammar of the Python language to a special program (the parser generator) which
+outputs the parser. The way the Python language is changed is therefore by
+modifiying the grammar file and developers rarely need to interact with the
+parser generator itself other than use it to generate the parser.
+
+How PEG Parsers Work
+--------------------
+
+.. _how-peg-parsers-work:
+
+A PEG (Parsing Expression Grammar) grammar differs from a context-free grammar
+(like the current one) in the fact that the way it is written more closely
+reflects how the parser will operate when parsing it. The fundamental technical
+difference is that the choice operator is ordered. This means that when writing::
+
+  rule: A | B | C
+
+a context-free-grammar parser (like an LL(1) parser) will generate constructions
+that given an input string will *deduce* which alternative (``A``, ``B`` or ``C``)
+must be expanded, while a PEG parser will check if the first alternative succeeds
+and only if it fails, will it continue with the second or the third one in the
+order in which they are written. This makes the choice operator not commutative.
```

Comment:
This is basically the same as order being important in ``and`` and ``or`` expressions and sequences of conditions in ``if`` and ``match`` statements.  Once I saw this, understanding the rest became easy.  So I think adding a sentence here would be useful.

Some theoretical computer scientists have advocated that multiple conditions be considered unordered like parsing patterns, so that multiple matches are possible, making a statement ambiguous.

Reply:
Yes and no. Grammars in general doesn't describe execution or matching. Indeed, EBNF grammars don't describe how to mach strings in the language but how to generate them. Algorithmic transformation is required to know how an EBNF grammar matches strings. PEG is unique here in the sense that th grammar is more related to the execution of the parser so what you mean is indeed closer, but care must be taken because it doesn't translate well to general grammars.